### PR TITLE
[ADD] DCOS-367 As a user, I want to scale the application

### PR DESCRIPTION
**Pull request by YuraTolochkevich**

### CODE REVIEW

**Comment by YuraTolochkevich** [comment]
@jsancio Hi Jose!  I have created new PR. please review

File: cli/dcoscli/marathon/main.py [hunk 1]
```diff
@@ -25,6 +25,7 @@
     dcos marathon group show [--group-version=<group-version>] <group-id>
     dcos marathon group remove [--force] <group-id>
     dcos marathon group update [--force] <group-id> [<properties>...]
+    dcos marathon group scale [--force] <group-id> <scale-factor>
```
Comment (jsancio) [inline cli/dcoscli/marathon/main.py]: Can we sort this line? It looks like it should go before `dcos marathon group show`

File: cli/dcoscli/marathon/main.py [hunk 2]
```diff
@@ -89,13 +90,13 @@
                                 (https://mesosphere.github.io/
                                 marathon/docs/rest-api.html#post-/v2/groups).
 
-    <instances>                 The number of instances to start
```
Comment (jsancio) [inline cli/dcoscli/marathon/main.py]: We still have this positional parameter: `dcos marathon app start ... [<instances>]`

File: cli/dcoscli/marathon/main.py [hunk 3]
```diff
@@ -645,6 +651,25 @@ def _update(app_id, properties, force):
     return 0
 
 
+def _group_scale(group_id, scale_factor, force):
+    """
+    :param group_id: the id of the group
+    :type group_id: str
+    :param scale_factor: scale factor for application group
+    :type properties: float
```
Comment (jsancio) [inline cli/dcoscli/marathon/main.py]: This should be `:type scale_factor: float`.

File: cli/tests/integrations/test_marathon_groups.py [hunk 4]
```diff
@@ -123,6 +123,22 @@ def test_update_missing_field():
             "Possible properties are: ")
 
 
+def test_scale_group():
```
Comment (jsancio) [inline cli/tests/integrations/test_marathon_groups.py]: This is great but let's also add negative tests. For example what happens when `<group-id>` doesn't exists, `<scale-factor>` is negative, `<scale-factor>` is not a flow (e.g. `dcos marathon group scale scale-group not_a_float`)

File: dcos/marathon.py [hunk 5]
```diff
@@ -388,6 +388,35 @@ def scale_app(self, app_id, instances, force=None):
...
+        if not force:
+            params = None
+        else:
+            params = {'force': 'true'}
+
+        url = self._create_url('v2/groups{}'.format(group_id))
+
+        response = http.put(url,
+                            params=params,
+                            json={'scaleBy': float(scale_factor)},
```
Comment (jsancio) [inline dcos/marathon.py]: Let's handle the error much better. I would pydoc for this method to expect a `float` and parse the input in the command: `cli/dcoscli/marathon/main.py`. We have code that parses `strings` at `int` and throws a `DCOSException` instead of a `ValueError`: https://github.com/mesosphere/dcos-cli/blob/master/dcos/util.py#L438-L454.

Let's do the same thing for `float` and add a function to the module `dcos.util` that parses as a float and throws a `DCOSException` if it fails.

File: cli/dcoscli/marathon/main.py (hunk 3 above)
Comment (jsancio) [inline cli/dcoscli/marathon/main.py]: Actually, `scale_factor` is not a flow but a `str`.

File: dcos/marathon.py [hunk 6]
```diff
@@ -388,6 +388,35 @@ def scale_app(self, app_id, instances, force=None):
         deployment = response.json()['deploymentId']
         return deployment
 
+    def scale_group(self, group_id, scale_factor, force=None):
+        """Scales a group with the requested scale-factor.
+        :param group_id: the ID of the group to scale
+        :type group_id: str
+        :param scale_factor: the requested value of scale-factor
+        :type scale_factor: int
```
Comment (jsancio) [inline dcos/marathon.py]: The type for `scale_factor` should be `float` and assume so in the code below. See my comment below.

**Comment by jsancio** [comment]
Thanks for the PR! Looks good in general. Left some comments. Please ping me when it is ready for my review.

**Comment by YuraTolochkevich** [comment]
@jsancio Thanks Jose!
Please see fresh changes!

File: dcos/marathon.py [hunk 7]
```diff
@@ -388,6 +388,37 @@ def scale_app(self, app_id, instances, force=None):
...
+        group_id = self.normalize_app_id(group_id)
+
+        if not force:
+            params = None
+        else:
+            params = {'force': 'true'}
+
+        url = self._create_url('v2/groups{}'.format(group_id))
+
+        scale_factor = util.parse_float(scale_factor)
```
Comment (jsancio) [inline dcos/marathon.py]: To match the rest of the code let's move this to `dcoscli.marathon.main.__group_scale`. This method `scale_group` should accept and assume that `scale_factor` is a `float`. Let's change the pydoc above to reflect that: `type scale_factor: float`.

File: cli/tests/integrations/test_marathon_groups.py [hunk 8]
```diff
@@ -123,6 +123,56 @@ def test_update_missing_field():
...
+
+    assert res['groups'][0]['apps'][0]['instances'] == 2
+    _remove_group('scale-group')
+
+
+def test_scale_group_not_exist():
+    returncode, stdout, stderr = exec_command(['dcos', 'marathon', 'group',
+                                               'scale', 'scale-group', '2'])
+    assert stderr == b''
+    watch_all_deployments()
```
Comment (jsancio) [inline cli/tests/integrations/test_marathon_groups.py]: Why is this here? It doesn't look like we are deploying anything.

File: cli/tests/integrations/test_marathon_groups.py [hunk 9]
```diff
@@ -123,6 +123,56 @@ def test_update_missing_field():
...
+    res = json.loads(stdout.decode('utf-8'))
+
+    assert res['groups'][0]['apps'][0]['instances'] == 2
+    _remove_group('scale-group')
+
+
+def test_scale_group_not_exist():
+    returncode, stdout, stderr = exec_command(['dcos', 'marathon', 'group',
+                                               'scale', 'scale-group', '2'])
+    assert stderr == b''
```
Comment (jsancio) [inline cli/tests/integrations/test_marathon_groups.py]: Shouldn't this print an error saying that `scale-group` doesn't exists?

File: cli/tests/integrations/test_marathon_groups.py [hunk 10]
```diff
@@ -123,6 +123,56 @@ def test_update_missing_field():
...
+
+    assert len(res['apps']) == 0
+    _remove_group('scale-group')
+
+
+def test_scale_group_when_scale_factor_negative():
+    _deploy_group('tests/data/marathon/groups/scale.json')
+    returncode, stdout, stderr = exec_command(['dcos', 'marathon', 'group',
+                                               'scale', 'scale-group', '-2'])
+    assert b'Command not recognized' in stdout
```
Comment (jsancio) [inline cli/tests/integrations/test_marathon_groups.py]: What is the full text that we are printing? Can you please provide a sample out for this use cases?

**Comment by jsancio** [comment]
@YuraTolochkevich Left you some comments.

File: cli/tests/integrations/test_marathon_groups.py (hunk 10 above)
Comment (YuraTolochkevich) [inline cli/tests/integrations/test_marathon_groups.py]: Docopt library recognize "-" symbol as command call with option "-2", which is not recognized.  Should we more accurately treat such situation?

File: cli/tests/integrations/test_marathon_groups.py (hunk 9 above)
Comment (YuraTolochkevich) [inline cli/tests/integrations/test_marathon_groups.py]: Marathon rest API works is such way, when we want to scale  group which doesn't exist, end send PUT request, new group with empty applications list is created, and status OK returned.
 For example:
curl -X PUT -i http://rslyusarc-ElasticL-QY77PH81BNNO-1491641693.us-west-2.elb.amazonaws.com/marathon/v2/groups/not-exist-group -d '{"scaleBy": 2.5 }'
HTTP/1.1 200 OK
Cache-Control: no-cache, no-store, must-revalidate
Content-Type: application/json; qs=2
Date: Wed, 26 Aug 2015 08:23:30 GMT
Expires: 0
Pragma: no-cache
Server: openresty/1.7.10.2
X-Marathon-Leader: http://$($MESOS_IP_DISCOVERY_COMMAND):8080
Content-Length: 92
Connection: keep-alive

{"version":"2015-08-26T08:23:30.119Z","deploymentId":"2f0cf4af-2c69-4d8e-8dae-71eb589fb3f6"}

curl -X GET -i http://rslyusarc-ElasticL-QY77PH81BNNO-1491641693.us-west-2.elb.amazonaws.com/marathon/v2/groups/not-exist-group 
HTTP/1.1 200 OK
Cache-Control: no-cache, no-store, must-revalidate
Content-Type: application/json; qs=2
Date: Wed, 26 Aug 2015 08:27:35 GMT
Expires: 0
Pragma: no-cache
Server: openresty/1.7.10.2
X-Marathon-Leader: http://$($MESOS_IP_DISCOVERY_COMMAND):8080
Content-Length: 102
Connection: keep-alive

{"id":"/not-exist-group","apps":[],"groups":[],"dependencies":[],"version":"2015-08-26T08:23:30.119Z"}

File: cli/tests/integrations/test_marathon_groups.py (hunk 8 above)
Comment (YuraTolochkevich) [inline cli/tests/integrations/test_marathon_groups.py]: Please see comment below

File: dcos/marathon.py (hunk 7 above)
Comment (YuraTolochkevich) [inline dcos/marathon.py]: OK

**Comment by YuraTolochkevich** [comment]
@jsancio Thanks Jose! Please take a look

File: cli/tests/integrations/test_marathon_groups.py (hunk 10 above)
Comment (jsancio) [inline cli/tests/integrations/test_marathon_groups.py]: Oh, right. No, that is fine.

**Comment by jsancio** [comment]
Perfect. LGTM!
